modules/safer_mysql/README.md

```diff
@@ -0,0 +1,250 @@
+# terraform-google-sql for a safer MySQL
+
+The module sets up a MySQL installation that relies on Cloud IAM for
+authenticating its users. The use of Cloud IAM centralises identty
+management, access control, and permits to strengthen the credentials
+being used for authentication.
+
+The module uses the terraform-google-sql module for MySQL, but controls
+more strictly the type of network connections that are allowed: in all
+cases, network connections need to be mediated via the Cloud SQL Proxy and,
+hence, be authorized via Cloud IAM.
+
+The most secure setup doesn't assign a public IP to the instance
+(assign_public_ip = false), and permits connections only from within the
+given VPC. In such a setup, sharing the instance between projects requires
+setting up Shared VPCs, and direct mysql access from engineering workstations
+is not possible: any debugging needs to be performed via bastion hosts or
+custom tooling running on VMs connected to the VPC.
+
+If assign_public_ip = true, the instance will be assigned a public IP. However,
```

Comment:
What is the use case for `assign_public_ip` on the safe module?

Reply:
The rationale is that there are certain networking setups that are hard to use with only private ips, and the mandated use of Cloud SQL Proxy for connections mitigates sufficient risks.

In particular, there are limits in the way the Cloud SQL peering works, as it doesn't allow transitive connections between peered networks. So if we have:

 P1 <--> Private Cloud SQL

and now we want to use the same instance from P2, one natural thing would be to peer its VPC with P1. However,  P2 won't be able to see Cloud SQL in this setup.

P2 <--> P1 <-X only P1 X-> Cloud SQL. 

The solution for this are shared VPCs, but adopting them often requires significant changes in the architecture, and there are opinions stating that Shared VPCs might end up eroding the security boundary defined by a project, and sharing too many resources by default.

Because we only expose a public IP with NO authorized networks, we ensure that Mysql is not directly exposed, but its connections are mediated by the Cloud SQL Proxy and authorized by Cloud IAM, making Cloud SQL to be similar to other Cloud APIs endpoints that are exposed to the public.

@yukinying did I miss anything?

Reply:
Got it, thanks for the thorough explanation.